sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageExtensions.cs

```diff
@@ -26,9 +27,21 @@ private static IEnumerable<Data> AsAmqpData(this IEnumerable<ReadOnlyMemory<byte
         {
             foreach (ReadOnlyMemory<byte> data in binaryData)
             {
+                ArraySegment<byte> segment;
```

Comment:
See https://github.com/Azure/azure-sdk-for-net/pull/19683#discussion_r598914753 for more details

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
AMD Ryzen 9 3950X, 1 CPU, 32 logical and 16 physical cores
.NET Core SDK=5.0.201
  [Host]     : .NET Core 3.1.13 (CoreCLR 4.700.21.11102, CoreFX 4.700.21.11602), X64 RyuJIT
  Job-QXDBJK : .NET Core 3.1.13 (CoreCLR 4.700.21.11102, CoreFX 4.700.21.11602), X64 RyuJIT

InvocationCount=320000  

```
|  Method | Elements |        Mean |    Error |   StdDev | Ratio | RatioSD |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|-------- |--------- |------------:|---------:|---------:|------:|--------:|-------:|-------:|------:|----------:|
| **ToArray** |        **1** |    **225.0 ns** |  **3.10 ns** |  **2.90 ns** |  **1.00** |    **0.00** | **0.0406** |      **-** |     **-** |     **344 B** |
| Marshal |        1 |    218.9 ns |  2.25 ns |  2.10 ns |  0.97 |    0.02 | 0.0344 |      - |     - |     304 B |
|         |          |             |          |          |       |         |        |        |       |           |
| **ToArray** |        **2** |    **386.2 ns** |  **2.70 ns** |  **2.39 ns** |  **1.00** |    **0.00** | **0.0594** |      **-** |     **-** |     **512 B** |
| Marshal |        2 |    396.9 ns |  3.45 ns |  3.23 ns |  1.03 |    0.01 | 0.0500 |      - |     - |     432 B |
|         |          |             |          |          |       |         |        |        |       |           |
| **ToArray** |        **4** |    **749.5 ns** | **14.98 ns** | **24.19 ns** |  **1.00** |    **0.00** | **0.1000** |      **-** |     **-** |     **848 B** |
| Marshal |        4 |    771.4 ns | 15.45 ns | 20.08 ns |  1.04 |    0.02 | 0.0813 |      - |     - |     688 B |
|         |          |             |          |          |       |         |        |        |       |           |
| **ToArray** |        **8** |  **1,515.2 ns** | **18.95 ns** | **16.80 ns** |  **1.00** |    **0.00** | **0.1813** |      **-** |     **-** |    **1520 B** |
| Marshal |        8 |  1,508.8 ns | 29.63 ns | 27.72 ns |  1.00 |    0.02 | 0.1406 |      - |     - |    1200 B |
|         |          |             |          |          |       |         |        |        |       |           |
| **ToArray** |       **16** |  **2,992.2 ns** | **40.45 ns** | **37.84 ns** |  **1.00** |    **0.00** | **0.3469** |      **-** |     **-** |    **2912 B** |
| Marshal |       16 |  2,990.0 ns | 37.82 ns | 35.38 ns |  1.00 |    0.01 | 0.2688 |      - |     - |    2272 B |
|         |          |             |          |          |       |         |        |        |       |           |
| **ToArray** |       **32** |  **5,852.9 ns** | **57.16 ns** | **53.47 ns** |  **1.00** |    **0.00** | **0.6844** | **0.0094** |     **-** |    **5728 B** |
| Marshal |       32 |  5,877.4 ns | 64.63 ns | 57.29 ns |  1.00 |    0.01 | 0.5313 | 0.0063 |     - |    4448 B |
|         |          |             |          |          |       |         |        |        |       |           |
| **ToArray** |       **64** | **11,681.2 ns** | **91.30 ns** | **85.40 ns** |  **1.00** |    **0.00** | **1.3563** | **0.0344** |     **-** |   **11360 B** |
| Marshal |       64 | 11,401.9 ns | 74.13 ns | 69.34 ns |  0.98 |    0.01 | 1.0500 | 0.0281 |     - |    8800 B |

Reply:
```
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Runtime.InteropServices;
    using System.Text;
    using BenchmarkDotNet.Attributes;
    using BenchmarkDotNet.Configs;
    using BenchmarkDotNet.Diagnosers;
    using BenchmarkDotNet.Exporters;
    using BenchmarkDotNet.Jobs;
    using Microsoft.Azure.Amqp;

    [Config(typeof(Config))]
    public class DataConversion
    {
        private IEnumerable<ReadOnlyMemory<byte>> data;

        private class Config : ManualConfig
        {
            public Config()
            {
                AddExporter(MarkdownExporter.GitHub);
                AddDiagnoser(MemoryDiagnoser.Default);
                AddJob(Job.Default.WithInvocationCount(320000));
            }
        }

        [Params(1, 2, 4, 8, 16, 32, 64)]
        public int Elements { get; set; }

        [IterationSetup]
        public void Setup()
        {
            data = Enumerable.Range(0, Elements)
                .Select(i => (ReadOnlyMemory<byte>) Encoding.UTF8.GetBytes($"Hello World {i}"));
        }

        [Benchmark(Baseline = true)]
        public ArraySegment<byte>[] ToArray()
        {
            return ConvertBefore(data);
        }

        static ArraySegment<byte>[] ConvertBefore(IEnumerable<ReadOnlyMemory<byte>> allData)
        {
            return allData.Select(data => new ArraySegment<byte>(data.IsEmpty ? Array.Empty<byte>() : data.ToArray())).ToArray();
        }

        [Benchmark]
        public ArraySegment<byte>[] Marshal()
        {
            return ConvertAfter(data);
        }

        static ArraySegment<byte>[] ConvertAfter(IEnumerable<ReadOnlyMemory<byte>> allData)
        {
            return allData.Select(data =>
            {
                ArraySegment<byte> segment;
                if (!data.IsEmpty)
                {
                    if (!MemoryMarshal.TryGetArray(data, out segment))
                    {
                        segment = new ArraySegment<byte>(data.ToArray());
                    }
                }
                else
                {
                    segment = new ArraySegment<byte>(Array.Empty<byte>());
                }

                return segment;
            }).ToArray();
        }
    }
```